### PR TITLE
[KYUUBI #6070] Improve perf on assembling row-based TRowSet

### LINKED ISSUE #6070: [Improvement] Performance Improvement for converting spark rows to thrift rows
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/kyuubi/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

https://issues.apache.org/jira/browse/SPARK-47085 reported an issue, scala Seq.apply has O(n) complexity so that we shall avoid it

### How should we improve?

_No response_

### Are you willing to submit PR?

- [ ] Yes. I would be willing to submit a PR with guidance from the Kyuubi community to improve.
- [X] No. I cannot submit a PR at this time.

### PR DESCRIPTION
**Pull request by Kwafoor**
# :mag: Description
## Issue References 🔗

This pull request fixes #6070

## Describe Your Solution 🔧

https://issues.apache.org/jira/browse/SPARK-47085

## Types of changes :bookmark:

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Related Unit Tests

---

# Checklist 📝

- [ ] This patch was not authored or co-authored using [Generative Tooling](https://www.apache.org/legal/generative-tooling.html)

**Be nice. Be informative.**

### CODE REVIEW

File: kyuubi-common/src/main/scala/org/apache/kyuubi/engine/result/TRowSetGenerator.scala [hunk 1]
```diff
@@ -65,13 +61,11 @@ trait TRowSetGenerator[SchemaT, RowT, ColumnT]
     val tRowSet = new TRowSet(0, new JArrayList[TRow](rowSize))
     var i = 0
     val columnSize = getColumnSizeFromSchemaType(schema)
-    val tColumns = new JArrayList[TColumn](columnSize)
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/engine/result/TRowSetGenerator.scala:68]: it's by design, see https://github.com/apache/kyuubi/pull/5831

Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/engine/result/TRowSetGenerator.scala]: ditto in row based implementation

Comment (Kwafoor) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/engine/result/TRowSetGenerator.scala]: sorry for mistake , thanks.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/engine/result/TRowSetGenerator.scala [hunk 2]
```diff
@@ -40,11 +42,7 @@ trait TRowSetGenerator[SchemaT, RowT, ColumnT]
   }
 
   def toRowBasedSet(rows: Seq[RowT], schema: SchemaT): TRowSet = {
-    val rowSize = rows.length
-    val tRows = new JArrayList[TRow](rowSize)
-    var i = 0
-    while (i < rowSize) {
-      val row = rows(i)
+    val tRows = rows.map { row =>
       var j = 0
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/engine/result/TRowSetGenerator.scala:46]: since there is no `i`, can we replace `j` with `i`?

**Review by pan3793 — APPROVED** [review]

**Review by cxzl25 — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks, merged to master

**Comment by yaooqinn** [comment]
TColumnGenerator.getColumnToList was missed?

**Comment by hh-cn** [comment]
@beryllw @yaooqinn 

I'm sorry to bother you, but it seems that this issue hasn't been fully resolved. The logic related to toRowBasedSet has been fixed, but in the toColumnBasedSet, the getColumnToList method still involves similar access to a non-IndexedSeq with val row = rows(idx). This results in significantly impacted serialization speed when Hive JDBC statements have a large fetchSize (> 300).

> trait TColumnGenerator[RowT] extends TRowSetColumnGetter[RowT] {
>   protected def getColumnToList[T](
>       rows: Seq[RowT],
>       ordinal: Int,
>       defaultVal: T,
>       convertFunc: (RowT, Int) => T = null): (JList[T], ByteBuffer) = {
>     val rowSize = rows.length
>     val ret = new JArrayList[T](rowSize)
>     val nulls = new JBitSet()
>     var idx = 0
>     while (idx < rowSize) {
>       val row = rows(idx)
>       val isNull = isColumnNullAt(row, ordinal)
>       if (isNull) {
>         nulls.set(idx, true)
>         ret.add(defaultVal)
>       } else {
>         val value = Option(convertFunc) match {
>           case Some(f) => f(row, ordinal)
>           case _ => getColumnAs[T](row, ordinal)
>         }
>         ret.add(value)
>       }
>       idx += 1
>     }
>     (ret, ByteBuffer.wrap(nulls.toByteArray))
>   }

**Comment by yaooqinn** [comment]
@hh-cn Can you send a PR for that issue?

**Comment by hh-cn** [comment]
> @hh-cn Can you send a PR for that issue?

I am not familiar with the process and standards for submitting a PR. I will glad to see you fixing it soon.

**Comment by yaooqinn** [comment]
Can you create an issue then? I'm fully booked and don't have time to fix this. But I'm sure someone will be interested in fixing it.

**Pull request closed without merging**